src/modules/job-manager/plugins/jobspec-default.c

```diff
@@ -0,0 +1,376 @@
+ * If still unset, set queue to NULL and return success.
+ * Return 0 on success.  On parse failure, set 'error' and return -1.
+ */
+static int get_queue (json_t *jobspec,
+                      json_t *defaults,
+                      const char **queue,
+                      flux_error_t *error)
+{
+    const char *s = NULL;
+    json_error_t e;
+
+    if (jobspec)
+        (void)json_unpack (jobspec, "{s?s}", "queue", &s);
+    if (!s && defaults) {
+        if (json_unpack_ex (defaults, &e, 0, "{s?s}", "queue", &s) < 0) {
+            errprintf (error, "%s", e.text);
+            return -1;
+        }
+    }
+    *queue = s;
+    return 0;
+}
+
+/* Create a new object consisting of key-values from o1 and o2,
+ * with o2 overwriting values from o1 when keys are present in both.
+ * The resulting object must be freed by the caller.
+ * N.B. object values are references on values in o1 and o2.
+ * Set to NULL if o1 and o2 are NULL and return success.
+ * Return 0 on success, -1 on failure.
+ */
+static int merge_tables (json_t *o1, json_t *o2, json_t **merged)
+{
+    json_t *o = NULL;
+
+    if (o1 || o2) {
+        if (!(o = json_object ()))
+            goto error;
+        if (o1 && json_object_update (o, o1) < 0)
+            goto error;
+        if (o2 && json_object_update (o, o2) < 0)
+            goto error;
+    }
+    *merged = o;
+    return 0;
+error:
+    json_decref (o);
+    return -1;
+}
+
+/* Catch any errors with the config early so that plugin load fails
+ * and someone can fix the config before jobs are submitted.
+ */
+static int validate_config (flux_t *h, flux_error_t *error)
+{
+    json_t *general_defaults;
+    json_t *queue_defaults;
+    const char *queue;
+    flux_error_t e;
+    json_t *queues = NULL;
+    json_t *value;
+
+    if (get_general_defaults (h, &general_defaults, &e) < 0) {
+        errprintf (error, "Error parsing default policy: %s", e.text);
+        return -1;
+    }
+    /* Check the default queue, which must exist in [queues].
+     */
+    if (get_queue (NULL, general_defaults, &queue, &e) < 0) {
+        errprintf (error, "Error parsing default queue name: %s", e.text);
+        return -1;
+    }
+    if (get_queue_defaults (h, queue, &queue_defaults, &e) < 0) {
+        errprintf (error, "Error parsing %s policy: %s", queue, e.text);
+        return -1;
+    }
+    /* Validate each member of [queues] also.
+     */
+    (void)flux_conf_unpack (flux_get_conf (h),
+                            NULL,
+                            "{s?o}",
+                            "queues", &queues);
+    if (queues) {
+        json_object_foreach (queues, queue, value) {
+            if (get_queue_defaults (h, queue, &queue_defaults, &e) < 0) {
+                errprintf (error,
+                           "Error parsing %s queue policy: %s",
+                           queue,
+                           e.text);
+                return -1;
+            }
+        }
+    }
```

Comment:
is this check redundant?  the foreach iterates over every value of `queues.X`.  `get_queue_defaults()` makes sure that `queues.X` exists.  Or am I missing something.

Reply:
`[queues]` is not required so that's why the foreach is in a block that tests for it being non-NULL.

Then each entry can optionally have `policy.jobspec.default.system` defined as a table, so we get that if available, and we throw an error if it cannot be parsed (e.g. what if `queue.pdebug.policy` had a value of 42 instead of a table?)